### PR TITLE
`treehouses uptime` (fixes #1250)

### LINKED ISSUE #1250: `treehouses uptime`
Gets the uptime of the RPi

### PR DESCRIPTION
**Pull request by law-lin**
Fixes #1250 

Displays the uptime or boot time of the Raspberry Pi
```
root@law-lin:~/cli# ./cli.sh uptime
 04:55:21 up  1:02,  0 users,  load average: 0.06, 0.14, 0.12
root@law-lin:~/cli# ./cli.sh uptime asd
Unknown operation provided.
root@law-lin:~/cli# ./cli.sh uptime boot
Raspberry Pi booted at:
2020-05-09 03:53:21
```

### CODE REVIEW

File: modules/uptime.sh [hunk 1]
```diff
@@ -0,0 +1,31 @@
+function uptime {
+  checkrpi
+  checkargn $# 1
+  if [[ $1 == "" ]]; then
+    if ! dpkg-query -W -f='${Status}' uptimed | grep -q "ok installed";
+    then
+        sudo apt-get install uptimed
```
Comment (kjong) [inline modules/uptime.sh:7]: You might want to move this package installation to builder

https://github.com/treehouses/builder/blob/23ad006b40d8e2365bce1d61c92e4a4f75411ead/scripts.d/16_pkg_install.sh#L5

File: modules/uptime.sh [hunk 2]
```diff
@@ -0,0 +1,31 @@
+function uptime {
+  checkrpi
+  checkargn $# 1
+  if [[ $1 == "" ]]; then
+    if ! dpkg-query -W -f='${Status}' uptimed | grep -q "ok installed";
```
Comment (kjong) [inline modules/uptime.sh:5]: We have a `check_missing_packages` in `globals.sh` that you can use.

**Comment by law-lin** [comment]
`treehouses uptime stop` and `treehouses uptime start` added. 

Use `systemctl --all` to check services running.

**Comment by law-lin** [comment]
root@law-lin:~/cli/tests# bats uptime.bats
 ✓  uptime
 ✓  uptime boot
 ✓  uptime start
 ✓  uptime stop

File: modules/help.sh [hunk 3]
```diff
@@ -90,14 +90,15 @@ Usage: treehouses
...
+   uptime [boot|start|stop]                  displays uptime and boot time of the Raspberry Pi           
    usb [on|off]                              turns usb ports on or off
    remote [check|status|upgrade|services]    helps with treehouses remote android app
           [version|commands|allservices]
    log <0|1|2|3|4|show|max>                  gets/sets log level and shows log
    blocker <0|1|2|3|4||max>                  website blocking levels using /etc/hosts
    sdbench                                   displays read and write speed of micro SD card
    inspire [joke|qotd|random]                displays quote based on user input
-   shutdown [now|in|force]                   shutdown the system 
+   shutdown [now|in|force]                   shutdown the system
```
Comment (ComputerOnFire) [inline modules/help.sh:101]: looks to be an extra space on the end of this line

File: modules/help.sh [hunk 4]
```diff
@@ -90,14 +90,15 @@ Usage: treehouses
    camera [on|off|capture]                   enables camera, disables camera, captures png photo
    cron [list|add|delete|deleteall]          adds, deletes a custom cron job or deletes, lists all cron jobs
         [0W|tor|timestamp]                   adds premade cron job (or removes it if already active)
+   uptime [boot|start|stop]                  displays uptime and boot time of the Raspberry Pi
```
Comment (ComputerOnFire) [inline modules/help.sh:93]: trailing spaces

File: modules/uptime.sh [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    command uptime -s
+  elif [ $1 == "stop" ]; then
+    systemctl stop uptimed
+  elif [ $1 == "start" ]; then
+    systemctl start uptimed
+  else
+    echo "Error: unknown operation provided"
+    uptime_help
+    exit 1
+  fi
```
Comment (ComputerOnFire) [inline modules/uptime.sh:19]: extra spaces on the end of fi

File: README.md [hunk 6]
```diff
@@ -125,7 +125,8 @@ log <0|1|2|3|4|show|max>                  gets/sets log level and shows log
 blocker <0|1|2|3|4||max>                  website blocking levels using /etc/hosts
 sdbench                                   displays read and write speed of micro SD card
 inspire [joke|qotd|random]                displays quote based on user input
-shutdown [now|in|force]                   shutdown the system           
+shutdown [now|in|force]                   shutdown the system   
+uptime [boot]                             displays uptime and boot time of the Raspberry Pi
```
Comment (dogi) [inline README.md:129]: please make the same as in help

**Comment by warthog16** [comment]
```
root@treehouses:~/cli/tests# bats uptime.bats 
 ✓  uptime
 ✗  uptime boot
   (from function `assert_success' in file /usr/lib/node_modules/bats-assert/src/assert.bash, line 114,
    in test file uptime.bats, line 11)
     `assert_success' failed
   
   -- command failed --
   status : 1
   output (15 lines):
     Error: unknown operation provided
     Usage: cli.sh uptime [boot|start|stop]
     
     Example:
       cli.sh uptime
           This returns the uptime of the Raspberry Pi
     
       cli.sh uptime boot
           This returns when the Raspberry Pi was booted
     
       cli.sh uptime start
           This starts Uptimed in the background
     
       cli.sh uptime stop
           This stops Uptimed from running in the background
   --
   
 ✗  uptime start
   (from function `assert_success' in file /usr/lib/node_modules/bats-assert/src/assert.bash, line 114,
    in test file uptime.bats, line 16)
     `assert_success' failed
   
   -- command failed --
   status : 1
   output (15 lines):
     Error: unknown operation provided
     Usage: cli.sh uptime [boot|start|stop]
     
     Example:
       cli.sh uptime
           This returns the uptime of the Raspberry Pi
     
       cli.sh uptime boot
           This returns when the Raspberry Pi was booted
     
       cli.sh uptime start
           This starts Uptimed in the background
     
       cli.sh uptime stop
           This stops Uptimed from running in the background
   --
   
 ✗  uptime stop
   (from function `assert_success' in file /usr/lib/node_modules/bats-assert/src/assert.bash, line 114,
    in test file uptime.bats, line 21)
     `assert_success' failed
   
   -- command failed --
   status : 1
   output (15 lines):
     Error: unknown operation provided
     Usage: cli.sh uptime [boot|start|stop]
     
     Example:
       cli.sh uptime
           This returns the uptime of the Raspberry Pi
     
       cli.sh uptime boot
           This returns when the Raspberry Pi was booted
     
       cli.sh uptime start
           This starts Uptimed in the background
     
       cli.sh uptime stop
           This stops Uptimed from running in the background
   --
   

4 tests, 3 failures
```
I think the testing file uses commands like `uptime current`, `uptime default`, and `uptime freq` which aren't defined.

**Comment by law-lin** [comment]
Test with this [image release](http://download.treehouses.io/experiment/treehouse-uptimed-fix2.img.gz), release-134 is no good.

**Comment by dogi** [comment]
@law-lin please fix merge conflict

**Pull request merged**